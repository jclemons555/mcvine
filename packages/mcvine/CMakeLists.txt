#
set(PACKAGE_NAME "mcvine")

set (LIB_NAME "${PACKAGE_NAME}")
set (PKG_NAME "${PACKAGE_NAME}")

#
add_subdirectory(python)
add_subdirectory(etc)
add_subdirectory(bin)
add_subdirectory(tests)
