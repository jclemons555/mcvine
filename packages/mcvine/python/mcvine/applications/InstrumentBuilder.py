# -*- Python -*-
# 
# Jiao Lin <jiao.lin@gmail.com>
#


'''
factory to create an instrumnt pyre application class
from a list of component names.
'''

import os


def build(neutron_components):
    
    class _Proxy:
        """proxy class of instrument sim app"""
        
        def __init__(self, *args, **kwds):
            self._init_params = args, kwds
            return

        def run(self, *args, **kwds):
            raise NotImplementedError
    
    _Proxy.neutron_components = neutron_components
    return _Proxy


def _build(neutron_components):
    
    from mcni.pyre_support.Instrument import Instrument as base
    class Instrument(base):

        class Inventory( base.Inventory ):

            from mcni.pyre_support import facility, componentfactory as component
            import mccomponents.pyre_support

            for name in neutron_components:
                code = '%s = facility("%s", default="mcni://optics/Dummy" )' % (
                    name, name)
                exec code in locals()
                continue
            del code, name

            import pyre.inventory as pinv
            # path of the post processing script
            # the components that need post-processing should append
            # to this script
            post_processing_script = pinv.str("post-processing-script")

            pass # end of Inventory


        def _defaults(self):
            base._defaults(self)
            self.inventory.sequence = neutron_components
            pps = self.inventory.post_processing_script
            if not pps:
                pps = os.path.join(self.inventory.outputdir, 'post-processing-script.py')
            self.post_processing_script = pps
            return


        def _makeSimContext(self):
            context = base._makeSimContext(self)
            context.post_processing_script = self.post_processing_script
            return context

        pass # end of Instrument

    return Instrument


# End of file 
