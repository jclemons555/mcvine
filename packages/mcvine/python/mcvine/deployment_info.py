# -*- Python -*-
#
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#
#                                   Jiao Lin
#                      California Institute of Technology
#                      (C) 2006-2011  All Rights Reserved
#
# {LicenseText}
#
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#


import os, sys
debug = os.environ.get('DEBUG', False)


# paths
dvdir = os.environ.get('DV_DIR')
exportroot = os.environ.get('EXPORT_ROOT')
pyredir = os.environ.get("PYRE_DIR")
mcvinedir = os.environ.get("MCVINE_DIR")
mcvine_resources = os.environ.get("MCVINE_RESOURCES")
mcvine_workflow = os.environ.get('MCVINE_WORKFLOW')

if dvdir and exportroot: type = 'developer'
else: 
    type = 'user'
    mcvinedir = mcvinedir or pyredir or exportroot
    if not mcvinedir:
        danse_dir = os.environ.get("DANSE_DIR")
        if not danse_dir:
            # if nothing is specified, assume that 
            # mcvine installed alongside the python
            # that is in use in the current user setting.
            opd = os.path.dirname
            danse_dir = opd(opd(sys.executable))
            msg = "Neither environment variable MCVINE_DIR nor DANSE_DIR was defined. Assumed that it is %s." % danse_dir
            import warnings
            warnings.warn(msg)
        mcvinedir = danse_dir

if not mcvine_resources and type == 'user':
    mcvine_resources = os.path.join(mcvinedir, 'share', 'mcvine', 'resources')
if mcvine_resources:
    mcvine_resources = os.path.abspath(mcvine_resources)

# default location of mcvine workflow
if not mcvine_workflow:
    mcvine_workflow = os.path.join(mcvinedir, 'share', 'mcvine', 'workflow')
sys.path.insert(0, mcvine_workflow)

# End of file 
