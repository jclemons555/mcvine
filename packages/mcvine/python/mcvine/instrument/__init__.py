# -*- Python -*-
#
# Jiao Lin <jiao.lin@gmail.com>
#

def main():
    import __main__ as m
    import sys
    assert len(sys.argv)==3
    ncount, buffer_size = map(int, sys.argv[1:])
    instrument = m.instrument
    N_iter = (ncount-1)//buffer_size + 1
    return

# End of file 
