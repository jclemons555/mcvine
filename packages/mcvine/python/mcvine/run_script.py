#
# Jiao Lin <jiao.lin@gmail.com>
#

import os, sys

def mpi_run(script, workdir, ncount, nodes, buffer_size=int(1e6), overwrite_datafiles=False):
    """run a mcvine simulation script on one node. The script must define the instrument.
Parameters:

* script: path to instrument script
* workdir: working dir
* ncount: neutron count
* nodes:  number of MPI nodes
* buffer_size: buffer size (optional)

An example script:

  # script myinstrument.py
  import mcvine
  instrument = mcvine.instrument()
  # add source
  source = mcvine.components.sources.Source_simple('source')
  instrument.append(source, position=(0,0,0))
  # add monitor
  monitor = mcvine.components.monitors.E_monitor('monitor', filename='IE.dat')
  instrument.append(monitor, position=(0,0,1))
  # end of script

"""
    return

def run1(script, workdir, ncount, buffer_size=int(1e6), overwrite_datafiles=False):
    """run a mcvine simulation script on one node. The script must define the instrument.

Parameters:

* script: path to instrument script
* workdir: working dir
* ncount: neutron count
* buffer_size: buffer size (optional)

An example script:

  # script myinstrument.py
  import mcvine
  instrument = mcvine.instrument()
  # add source
  source = mcvine.components.sources.Source_simple('source')
  instrument.append(source, position=(0,0,0))
  # add monitor
  monitor = mcvine.components.monitors.E_monitor('monitor', filename='IE.dat')
  instrument.append(monitor, position=(0,0,1))
  # end of script

"""
    import imp
    m = imp.load_source('mcvinesim', script)
    assert hasattr(m, 'instrument')
    instrument = m.instrument
    ncount = int(ncount)
    workdir = os.path.abspath(workdir)
    ppsd = os.path.join(workdir, 'post-processing-scripts')
    if not os.path.exists(ppsd): os.makedirs(ppsd)
    
    N = (ncount-1)//buffer_size+1
    remained = ncount
    for i in range(N):
        n = min(remained, buffer_size)
        instrument.simulate(n, outputdir=workdir, overwrite_datafiles=overwrite_datafiles, iteration_no=i, post_processing_scripts_dir=ppsd)
        remained -= n
        continue
    assert remained == 0
    for comp in instrument.components:
        comp.create_pps()
        continue
    from mcni import run_ppsd
    run_ppsd(ppsd)
    return


import click
@click.command()
@click.argument("script")
@click.option("--workdir", default="output", help="work directory")
@click.option("--ncount", default=int(1e6), help="neutron count")
@click.option("--overwrite_datafiles", default=False, help="overwrite datafiles", is_flag=True)
@click.option("--buffer_size", default=int(1e6), help="neutron buffer size")
def main(script, workdir, ncount, buffer_size=int(1e6), overwrite_datafiles=False):
    run1(script, workdir, ncount, buffer_size, overwrite_datafiles)
    return

if __name__ == '__main__': main()
