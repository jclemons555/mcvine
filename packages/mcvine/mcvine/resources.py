# -*- Python -*-
# Jiao Lin <linjiao@caltech.edu>


def root():
    from .deployment_info import mcvine_resources
    if not mcvine_resources:
        alert_missing_resources()
    return mcvine_resources


def alert_missing_resources():
    msg = """Missing information about MCViNE resources. 

If it is available in your system, please set environment
variable MCVINE_RESOURCES to its root path.

If it is not available, please obtain it by

 $ git clone https://github.com/heetuu/mcvine-resources.git

, and then set MCVINE_RESOURCES to the path of the new
directory.

"""
    raise IOError(msg)


import os

def instrument(name):
    r = root()
    return os.path.join(r, "instruments", name)


# End of file 

