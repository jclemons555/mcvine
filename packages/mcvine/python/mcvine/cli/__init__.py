# -*- Python -*-
#
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#
#                                   Jiao Lin
#                      California Institute of Technology
#                      (C) 2006-2013  All Rights Reserved
#
# {LicenseText}
#
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#


__doc__ = """
command line interface
"""

import click

# map aliases to long commands
aliases = dict()

# main command
# enable -h seems to interfere with the pyre_app decorator.
# disable it for now
# CONTEXT_SETTINGS = dict(help_option_names=['-h', '--help'])
# @click.group(context_settings=CONTEXT_SETTINGS)
@click.group()
def mcvine():
    return

# decorator to create bash alias of a command
def alias(shortname, longname):
    def decorate(f):
        aliases[shortname] = longname
        return f
    return decorate

# decorator for pyre app
def pyre_app(parent, appname, cmd_prefix):
    def decorator(f):
        # pass extra args and options to pyre app
        d1 = parent.command(
            context_settings=dict(ignore_unknown_options=True, allow_extra_args=True))
        d2 = click.pass_context
        def _f(ctx):
            # build the sys argv list for pyre app
            import sys
            sys.argv = [appname] + ctx.args
            # create app instance
            app = f(ctx, appname)
            # and run
            app.run()
            return
        # nicer cmd name
        _f.__name__ = f.__name__
        _f.__doc__ = f.__doc__
        # register the alias
        # sth like arcs_analyze_beam -> mcvine instrument arcs analyze_beam
        aliases[appname] = '%s %s' % (cmd_prefix, f.__name__)
        return d1(d2(_f))
    return decorator

# sub-cmds
from . import mpi, sampleassembly, mantid #, kernel
from mcvine.instrument import cli
from mcvine.instruments import cli
from mcstas2 import cli

# aliases should be the last cmds to import
from . import bash

# version
__id__ = "$Id$"

# End of file 
