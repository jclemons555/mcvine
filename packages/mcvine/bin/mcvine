#!/usr/bin/env python
#
# Jiao Lin <jiao.lin@gmail.com>
#

from mcvine.cli import main


if __name__ == '__main__': 
    print '-'*78
    main()


# End of file 
