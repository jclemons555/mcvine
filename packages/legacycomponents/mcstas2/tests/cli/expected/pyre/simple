#!/usr/bin/env python
import warnings
warnings.simplefilter('ignore')
import mcvine
warnings.simplefilter('default')

def main():
    from mcvine.applications.InstrumentBuilder import build
    components = ['arm', 'source', 'collimator', 'target', 'arm2', 'PSD_4pi']
    App = build(components)
    app = App('simple')
    app.run()
    return

if __name__ == '__main__': main()

# This application was created by the following command:
# $ mcvine-create-instrument-simulation-application -name=simple -components=arm,source,collimator,target,arm2,PSD_4pi

