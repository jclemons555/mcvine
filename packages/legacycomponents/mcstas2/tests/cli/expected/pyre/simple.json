[{"position": [[0.0, 0.0, 0.0], "absolute", null], "type": "Arm", "orientation": [[0, 0, 0], "absolute", null], "parameters": {}, "name": "arm"}, {"position": [[0.0, 0.0, 0.0], "relative", "arm"], "type": "Source_simple", "orientation": [[0, 0, 0], "relative", "arm"], "parameters": {"yh": "0.015", "dist": "1", "dE": "0.2", "xw": "0.024", "radius": "0.015", "E0": "5"}, "name": "source"}, {"position": [[0.0, 0.0, 0.4], "relative", "arm"], "type": "Collimator_linear", "orientation": [[0, 0, 0], "relative", "arm"], "parameters": {"ymax": "0.03", "divergence": "coll_div", "len": "0.2", "xmax": "0.02", "xmin": "-0.02", "ymin": "-0.03"}, "name": "collimator"}, {"position": [[0.0, 0.0, 1.0], "relative", "arm"], "type": "V_sample", "orientation": [[0, 0, 0], "relative", "arm"], "parameters": {"focus_r": "0", "yheight": "0.015", "target_z": "1", "target_x": "0", "target_y": "0", "xwidth": "0.02", "zthick": "0.004", "pack": "1"}, "name": "target"}, {"position": [[0.0, 0.0, 0.0], "relative", "target"], "type": "Arm", "orientation": [[0.0, "ROT", 0.0], "relative", "arm"], "parameters": {}, "name": "arm2"}, {"position": [[0.0, 0.0, 0.0], "relative", "arm2"], "type": "PSD_monitor_4PI", "orientation": [[0, 0, 0], "relative", "arm2"], "parameters": {"nx": "101", "ny": "51", "radius": "10", "filename": "\"vanadium.psd\""}, "name": "PSD_4pi"}]